python/ray/data/dataset.py

```diff
@@ -2264,6 +2264,7 @@ def iter_batches(
             blocks.iter_blocks(),
             stats,
             prefetch_blocks=prefetch_blocks,
+            clear_block_after_read=self._lazy,
```

Comment:
The lazy ones are expected to compute the blocks on the fly, whereas eager ones are expected to have materialized blocks. For other places, the reason was the same.
Now I reverted changes to Dataset and make this PR focus on DatasetPipeline since the API semantics for DatasetPipeline make it very clear about the safety (always lazy + read at most once).